### ISSUE TITLE
[HOLD for payment 2023-08-02] [HOLD for payment 2023-07-24] [$1000] Web - Pasting text or link in edit message pastes the text or link in main compose box

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#22803]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
. Open the app
2. Open any report
3. Send any message
4. Copy any text
5. Edit message sent in step 3
6. Paste the text in edit message and observe that text is pasted in compose box

## Expected Result:
App should paste text in edit message if edit message is in focus

## Actual Result:
App pastes text in main compose box even if edit message is in focus

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.40-1
**Reproducible in staging?:** y
**Reproducible in production?:** n
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/93399543/e27772be-d689-4050-8c3f-5393e9e30d60

https://github.com/Expensify/App/assets/93399543/756e2dc9-c7eb-4a41-96ee-dc864babd847

**Expensify/Expensify Issue URL:**
**Issue reported by:** GH handle - @dhanashree-sawant
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1689249016488049

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~019d975b520ddadde8</li>
        <li>Upwork Job ID: 1679978909772632064</li>
        <li>Last Price Increase: 2023-07-14</li>
    </ul>
</details>

**Comment by situchan** [comment]
Might be related to https://github.com/Expensify/App/pull/21583

**Comment by robertKozik** [comment]
Yeah, I'm looking on it. Probably have the fix, just need to double check it

**Comment by robertKozik** [comment]
I've created PR with fix: https://github.com/Expensify/App/pull/22817

**Comment by Santhosh-Sellavel** [comment]
@robertKozik this is the problem at hand now

Revert fixes  ✅   https://github.com/Expensify/App/issues/22833

**Comment by Santhosh-Sellavel** [comment]
@robertKozik Any update?

**Comment by Santhosh-Sellavel** [comment]
@cristipaval or @francoisl remove Awaiting Payment labels and add back the external label here!

**Comment by b4s36t4** [comment]
@Santhosh-Sellavel @francoisl if this is up for the External here's is my Proposal.

## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Web - Pasting text or link in edit message pastes the text or link in main compose box 
### What is the root cause of that problem?

The RCA for this issue is how we're handling the paste event listener. For every composer component we're creating an eventListener for `paste` event which is results in multiple event listeners. Also since the first event listener is added for main composer this causing the issue that even we paste text into the edit message composer main composer getting focus because of line here.

https://github.com/Expensify/App/blob/b9508b616f3a7229365c8bd7280fd1e34cfd2f9b/src/components/Composer/index.js#L281

### What changes do you think we should make in order to solve the problem?

> This change is only web I think cause ios and android have different logic.

We can't directly have `paste` listener on input, we can have listeners to added once an input get focus but this won't allow copy pasting files when the input is not focused.

We can overcome this issue using the event. Since the paste event will sends the `target` of the event where it occurred we can use that target to focus and run the code.

We need to create a new ref maybe `focusedInputRef` and assign to this inside `handlePaste` function.

```jsx
this.focusedInputRef.current = event.target
``` 
here
https://github.com/Expensify/App/blob/b9508b616f3a7229365c8bd7280fd1e34cfd2f9b/src/components/Composer/index.js#L309

And here https://github.com/Expensify/App/blob/b9508b616f3a7229365c8bd7280fd1e34cfd2f9b/src/components/Composer/index.js#L280 use the `focusedInputRef` to `focus` or `blur`. Since `execCommand` works on the field which is focused, this works pretty well with the text or HTML or even emojis but for files the functionality would be still same.. once we click it would be triggered by main composer only not from the one which we're editing (I don't know how to fix this :(, maybe this is expected or so.).



### What alternative solutions did you explore? (Optional)
NA

## Result
[report.webm](https://github.com/Expensify/App/assets/59088937/e380230c-b4ec-4848-ae15-964cedbfa96f)

Willing to evlaute my solution against multiple test-cases to be future proof.

**Comment by s-alves10** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Pasting text or link in editing message pastes the text or link in the main compose box

### What is the root cause of that problem?

We're using the same component in main composer and editing message: [Composer](https://github.com/Expensify/App/blob/main/src/components/Composer/index.js)
When pasting, we focuses the input and paste the content to it using document's native function.

https://github.com/Expensify/App/blob/6e5b07cf45944f28373e188d6f02620a12d6da29/src/components/Composer/index.js#L279-L290

We add document paste event handlers for all editing messages and main composer. And in my observation, only the latest added paste event handler is called. In our case, the main composer paste event handler is called because the main composer is mounted at the last among composers.

This is the root cause

### What changes do you think we should make in order to solve the problem?

`paste` event should be called only on focused element. Now we are adding event handlers on composer mount or navigation focus.
I think it is reasonable to add paste event handler on `focus` event of the input element and remove handler on `blur` event of the input element.

Replace

https://github.com/Expensify/App/blob/6e5b07cf45944f28373e188d6f02620a12d6da29/src/components/Composer/index.js#L171-L177

with
```
        this.textInput.addEventListener('focus', () => document.addEventListener('paste', this.handlePaste));
        this.textInput.addEventListener('blur', () => document.removeEventListener('paste', this.handlePaste));
```

And below code isn't needed anymore. Remove them

https://github.com/Expensify/App/blob/6e5b07cf45944f28373e188d6f02620a12d6da29/src/components/Composer/index.js#L210-L212

This works as expected.

**Note**: I think we don't need below codes
https://github.com/Expensify/App/blob/6e5b07cf45944f28373e188d6f02620a12d6da29/src/components/Composer/index.js#L281
https://github.com/Expensify/App/blob/6e5b07cf45944f28373e188d6f02620a12d6da29/src/components/Composer/index.js#L286-L287

### What alternative solutions did you explore? (Optional)

**Comment by robertKozik** [comment]
@Santhosh-Sellavel Sorry I was OOO already when you commented. I saw that `external` and `help wanted` labels were added, so should I handle this issue, or rather help you with proposal checking?

**Comment by Santhosh-Sellavel** [comment]
@robertKozik It's better we take care of fixing it here. Can you come up with a robust solution for this?

**Comment by robertKozik** [comment]
Hi @Santhosh-Sellavel I was tweaking around with that and I think I have one proposition, it actually consists from two already proposed snippets/ideas.
1) snippet from [here](https://github.com/Expensify/App/pull/22817#discussion_r1263015766) to fix everything except the cursor issue. I believe it's the most elegant fix in this case, as we're not creating additional listeners to target only one particular textInput. Additionally, we won't be splitting the handling of pasting inside and outside of the input.
2) Fix the cursor problem - with that im not sure whether this should be fixed alongside this issue, as it looks like the regression from another PR https://github.com/Expensify/App/pull/22443, which is stated [here](https://github.com/Expensify/App/issues/22833#issuecomment-1635174418). If we should include this I'll go with the solution from [this insight](https://github.com/Expensify/App/issues/22833#issuecomment-1635174418) too.

Let me know what you think about this approach. Aside from that, I experimented with multiple listeners, one on the document for outside pasting and another for the textInput itself, but it feels overengineered.

**Comment by Santhosh-Sellavel** [comment]
Thanks, @robertKozik 

Ya, the snippet looks good to me as well, please ensure it doesn't break anything!

About the cursor issue, it's better to revert and let them work on it. 

cc: @francoisl @cristipaval

**Comment by robertKozik** [comment]
@Santhosh-Sellavel I created draft PR with follow-up https://github.com/Expensify/App/pull/23135 . Should be wait for reverting the cursor bug?

**Comment by s-alves10** [comment]
@Santhosh-Sellavel 

What do you think about my [proposal](https://github.com/Expensify/App/issues/22803#issuecomment-1637341116)?

**Comment by robertKozik** [comment]
@Santhosh-Sellavel Can I make my PR `ready for review` or are we waiting for reverting the cursor bug?

**Comment by Santhosh-Sellavel** [comment]
@robertKozik  please make it ready for review.

**Comment by Santhosh-Sellavel** [comment]
@robertKozik Bump!

**Comment by robertKozik** [comment]
@Santhosh-Sellavel Today PR will be ready

**Comment by cristipaval** [comment]
It seems that the payment is due here.

**Comment by zanyrenney** [comment]
Payment on its way, I am working on this now @cristipaval - I was OOO before.

**Comment by zanyrenney** [comment]
Hmm, there is no accompanying upwork job. The only payment on Upwork also is the issue reporter, as per the comment above @cristipaval 👍🏼 

@robertKozik does not require payment (Contractor)
@Santhosh-Sellavel does not require payment (Eligable for Manual Requests)

**Comment by zanyrenney** [comment]
Upwork isn't letting me create a new job, keep getting an error. I will do this ASAP but looking like tomorrow so I can do a full clear cache, logout.

**Comment by Santhosh-Sellavel** [comment]
No Payments are due here for Myself & @robertKozik  as it is a regression handling. 

I think we can skip the checklist here as well.
All parties are aware of the mistake, and these tests will decide on the origin issue.

**Comment by cristipaval** [comment]
Thanks @Santhosh-Sellavel! 
@zanyrenney the only due payment is the reporting bonus for @dhanashree-sawant!

**Comment by zanyrenney** [comment]
Still waiting on @dhanashree-sawant to accept job offer and finish payout - link is here https://www.upwork.com/jobs/~019f97a860cbdfce94

**Comment by dhanashree-sawant** [comment]
Hi @zanyrenney, when I click on the job link, it is giving me 'Access denied' error

**Comment by dhanashree-sawant** [comment]
Hi @zanyrenney, I have just recieved the invite but job is kept hourly, can you change that to fixed price?

**Comment by zanyrenney** [comment]
new upwork job - https://www.upwork.com/jobs/~011e7524a423ae1245

**Comment by zanyrenney** [comment]
reporting bonus paid. 

all other payments handled via NewDot. closing!
